h2o-mapreduce-generic/src/main/java/water/hadoop/FileSystemBasedClouding.java

```diff
@@ -0,0 +1,140 @@
+package water.hadoop;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.*;
+
+import java.io.*;
+import java.net.InetAddress;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+
+public class FileSystemBasedClouding extends AbstractClouding {
+
+  private InetAddress _self_ip;
+  private int _self_port = -1;
+  
+  private int _n;
+  private Path _path;
+  private FileSystem _fs;
+
+  private final Map<String, URI> _nodes = new HashMap<>();
+
+  void init(Configuration conf) throws IOException {
+    _n = conf.getInt(h2omapper.H2O_CLOUD_SIZE_KEY, -1);
+    _path = new Path(conf.get(h2omapper.H2O_CLOUDING_DIR_KEY));
+    _fs = FileSystem.get(conf);
+  }
+
+  @Override
+  public void notifyAboutCloudSize(InetAddress ip, int port, InetAddress leaderIp, int leaderPort, int size) {
+    if (size != _n)
+      return;
+    Path path = toNodePath(ip, port, "leader");
+    try {
+      writeFile(path, h2oUri(leaderIp, leaderPort).toString());
+    } catch (IOException e) {
+      e.printStackTrace();
+      exit(162);
+    }
+  }
+
+  @Override
+  public void notifyAboutEmbeddedWebServerIpPort(InetAddress ip, int port) {
+    _self_ip = ip;
+    _self_port = port;
+
+    Path nodePath = toNodePath(ip, port, "node");
+    try {
+      writeFile(nodePath, h2oUri(ip, port).toString());
+    } catch (IOException e) {
+      e.printStackTrace();
+      exit(160);
+    }
+  }
+
+  @Override
+  public boolean providesFlatfile() {
+    return true;
+  }
+
+  @Override
+  public String fetchFlatfile() throws Exception {
+    while (_nodes.size() < _n) {
+      for (FileStatus fs : _fs.listStatus(_path, new NewNodesPathFilter())) {
+        String name = fs.getPath().getName();
+        if (name.endsWith(".exit")) {
+          exit(161);
+          return null;
+        }
+        try (InputStream is = _fs.open(fs.getPath());
+             BufferedReader br = new BufferedReader(new InputStreamReader(is))) {
+          String node = br.readLine();
+          _nodes.put(name, URI.create(node));
+        }
+      }
+      try {
+        Thread.sleep(1000);
```

Comment:
the correct call here is wait(), bussy waiting is not good
also InterruptedException should never be eaten, the best way to handle it to do         Thread.currentThread().interrupt();

Reply:
Agreed about the handling of the interrupt - I will just propagate it to the caller (who will shut down h2o immediately).

Regarding `wait` vs `sleep`, imho it doesn't matter here - wait() could make the whole thing more complex due to spurious wakeups: https://hazelcast.com/blog/spurious-wakeups-are-real-4/